docker/Dockerfile-pyston

```diff
@@ -0,0 +1,135 @@
+# syntax=docker/dockerfile:1
+# Dockerfile to build the matrixdotorg/synapse docker images using pyston.
+#
+# Note that it uses features which are only available in BuildKit - see
+# https://docs.docker.com/go/buildkit/ for more information.
+#
+# To build the image, run `docker build` command from the root of the
+# synapse repository:
+#
+#    DOCKER_BUILDKIT=1 docker build -f docker/Dockerfile-pyston .
+#
+
+# Irritatingly, there is no blessed guide on how to distribute an application with its
+# poetry-managed environment in a docker image. We have opted for
+# `poetry export | pip install -r /dev/stdin`, but there are known bugs in
+# in `poetry export` whose fixes (scheduled for poetry 1.2) have yet to be released.
+# In case we get bitten by those bugs in the future, the recommendations here might
+# be useful:
+#     https://github.com/python-poetry/poetry/discussions/1879#discussioncomment-216865
+#     https://stackoverflow.com/questions/53835198/integrating-python-poetry-with-docker?answertab=scoredesc
+
+
+
+###
+### Stage 0: generate requirements.txt
+###
+FROM pyston/pyston:2.3.3 as requirements
+
+# RUN --mount is specific to buildkit and is documented at
+# https://github.com/moby/buildkit/blob/master/frontend/dockerfile/docs/syntax.md#build-mounts-run---mount.
+# Here we use it to set up a cache for apt (and below for pip), to improve
+# rebuild speeds on slow connections.
+RUN \
+   --mount=type=cache,target=/var/cache/apt,sharing=locked \
+   --mount=type=cache,target=/var/lib/apt,sharing=locked \
+ apt-get update && apt-get install -y git rustc \
+    && rm -rf /var/lib/apt/lists/*
```

Comment:
This is one of the differences with the current image. `poetry` [requires cryptography](https://github.com/python-poetry/poetry/blob/34c6da4c2cbdfd5572fc4711d3ee628bf80eba72/poetry.lock#L133-L139), which doesn't publish binary wheels for pyston, hence needing `rustc`. 

From below:

```
# We install poetry in its own build stage to avoid its dependencies conflicting with
# synapse's dependencies.
```

I wonder if instead we should just install poetry to a different python prefix (`--prefix="/poetry"`), thus avoiding conflicts while also reducing the need to install native dependencies twice?